File: frontend/web-app/src/types/report.ts
export interface Report {
  report_id: string;
  report_name: string;
  report_type: 'fund_performance' | 'investor_statement' | 'capital_call' | 'distribution' | 'tax_report' | 'compliance' | 'custom';
  description?: string;
  fund_id?: string;
  investor_id?: string;
  period_start?: string;
  period_end?: string;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  created_by: string;
  created_at: string;
  completed_at?: string;
  file_path?: string;
  parameters: ReportParameters;
}

export interface ReportParameters {
  [key: string]: any;
  include_charts?: boolean;
  currency?: string;
  date_format?: string;
  grouping?: 'fund' | 'investor' | 'quarter' | 'year';
  filters?: ReportFilters;
}

export interface ReportFilters {
  fund_ids?: string[];
  investor_ids?: string[];
  date_range?: {
    start: string;
    end: string;
  };
  status?: string[];
  minimum_amount?: number;
  maximum_amount?: number;
}

export interface ReportTemplate {
  template_id: string;
  template_name: string;
  report_type: Report['report_type'];
  description?: string;
  default_parameters: ReportParameters;
  is_system: boolean;
  created_by: string;
  created_at: string;
  updated_at: string;
}

export interface DashboardMetrics {
  total_funds: number;
  total_investors: number;
  total_aum: number;
  total_committed: number;
  total_called: number;
  total_distributed: number;
  avg_fund_performance: number;
  pending_capital_calls: number;
  pending_distributions: number;
  recent_activity: ActivityItem[];
}

export interface ActivityItem {
  activity_id: string;
  activity_type: 'capital_call' | 'distribution' | 'new_investor' | 'fund_creation' | 'report_generated';
  description: string;
  timestamp: string;
  fund_id?: string;
  investor_id?: string;
  amount?: number;
  status?: string;
}

export interface PerformanceMetrics {
  fund_id: string;
  fund_name: string;
  as_of_date: string;
  nav: number;
  irr: number;
  moic: number;
  tvpi: number;
  dpi: number;
  rvpi: number;
  called_capital: number;
  distributed_capital: number;
  remaining_value: number;
  quarterly_returns: QuarterlyReturn[];
  benchmark_comparison?: BenchmarkComparison;
}

export interface QuarterlyReturn {
  quarter: string;
  year: number;
  return_percentage: number;
  nav: number;
  capital_calls: number;
  distributions: number;
}

export interface BenchmarkComparison {
  benchmark_name: string;
  benchmark_return: number;
  fund_return: number;
  outperformance: number;
  tracking_error: number;
  sharpe_ratio: number;
}

export interface CreateReportRequest {
  report_name: string;
  report_type: Report['report_type'];
  description?: string;
  fund_id?: string;
  investor_id?: string;
  period_start?: string;
  period_end?: string;
  parameters: ReportParameters;
}

export interface ReportSchedule {
  schedule_id: string;
  report_template_id: string;
  schedule_name: string;
  frequency: 'daily' | 'weekly' | 'monthly' | 'quarterly' | 'annually';
  next_run: string;
  is_active: boolean;
  recipients: string[];
  created_at: string;
  updated_at: string;
}

File: backend/api-gateway/src/models/report.py
from datetime import datetime
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field
from enum import Enum


class ReportType(str, Enum):
    FUND_PERFORMANCE = "fund_performance"
    INVESTOR_STATEMENT = "investor_statement"
    CAPITAL_CALL = "capital_call"
    DISTRIBUTION = "distribution"
    TAX_REPORT = "tax_report"
    COMPLIANCE = "compliance"
    CUSTOM = "custom"


class ReportStatus(str, Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class Frequency(str, Enum):
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"
    QUARTERLY = "quarterly"
    ANNUALLY = "annually"


class ActivityType(str, Enum):
    CAPITAL_CALL = "capital_call"
    DISTRIBUTION = "distribution"
    NEW_INVESTOR = "new_investor"
    FUND_CREATION = "fund_creation"
    REPORT_GENERATED = "report_generated"


class ReportFilters(BaseModel):
    fund_ids: Optional[List[str]] = None
    investor_ids: Optional[List[str]] = None
    date_range: Optional[Dict[str, str]] = None
    status: Optional[List[str]] = None
    minimum_amount: Optional[float] = None
    maximum_amount: Optional[float] = None


class ReportParameters(BaseModel):
    include_charts: Optional[bool] = True
    currency: Optional[str] = "USD"
    date_format: Optional[str] = "YYYY-MM-DD"
    grouping: Optional[str] = None
    filters: Optional[ReportFilters] = None


class Report(BaseModel):
    report_id: str
    report_name: str
    report_type: ReportType
    description: Optional[str] = None
    fund_id: Optional[str] = None
    investor_id: Optional[str] = None
    period_start: Optional[datetime] = None
    period_end: Optional[datetime] = None
    status: ReportStatus
    created_by: str
    created_at: datetime
    completed_at: Optional[datetime] = None
    file_path: Optional[str] = None
    parameters: ReportParameters


class ReportTemplate(BaseModel):
    template_id: str
    template_name: str
    report_type: ReportType
    description: Optional[str] = None
    default_parameters: ReportParameters
    is_system: bool = False
    created_by: str
    created_at: datetime
    updated_at: datetime


class ActivityItem(BaseModel):
    activity_id: str
    activity_type: ActivityType
    description: str
    timestamp: datetime
    fund_id: Optional[str] = None
    investor_id: Optional[str] = None
    amount: Optional[float] = None
    status: Optional[str] = None


class DashboardMetrics(BaseModel):
    total_funds: int
    total_investors: int
    total_aum: float
    total_committed: float
    total_called: float
    total_distributed: float
    avg_fund_performance: float
    pending_capital_calls: int
    pending_distributions: int
    recent_activity: List[ActivityItem] = []


class QuarterlyReturn(BaseModel):
    quarter: str
    year: int
    return_percentage: float
    nav: float
    capital_calls: float
    distributions: float


class BenchmarkComparison(BaseModel):
    benchmark_name: str
    benchmark_return: float
    fund_return: float
    outperformance: float
    tracking_error: float
    sharpe_ratio: float


class PerformanceMetrics(BaseModel):
    fund_id: str
    fund_name: str
    as_of_date: datetime
    nav: float
    irr: float
    moic: float
    tvpi: float
    dpi: float
    rvpi: float
    called_capital: float
    distributed_capital: float
    remaining_value: float
    quarterly_returns: List[QuarterlyReturn] = []
    benchmark_comparison: Optional[BenchmarkComparison] = None


class CreateReportRequest(BaseModel):
    report_name: str = Field(..., min_length=2, max_length=100)
    report_type: ReportType
    description: Optional[str] = None
    fund_id: Optional[str] = None
    investor_id: Optional[str] = None
    period_start: Optional[datetime] = None
    period_end: Optional[datetime] = None
    parameters: ReportParameters


class ReportSchedule(BaseModel):
    schedule_id: str
    report_template_id: str
    schedule_name: str
    frequency: Frequency
    next_run: datetime
    is_active: bool = True
    recipients: List[str] = []
    created_at: datetime
    updated_at: datetime